python/lsst/daf/butler/_exceptions.py

```diff
@@ -61,3 +114,30 @@ def __init__(self, reasons: list[str]):
     def __str__(self) -> str:
         # There may be multiple reasons, format them into multiple lines.
         return "Possible reasons for empty result:\n" + "\n".join(self.reasons)
+
+
+class UnknownButlerUserError(ButlerUserError):
+    """Raised when the server sends an ``error_type`` for which we don't know
+    the corresponding exception type.  (This may happen if an old version of
+    the Butler client library connects to a new server).
+    """
+
+    error_type = "unknown"
+
+
+_USER_ERROR_TYPES: tuple[type[ButlerUserError], ...] = (
+    DatasetNotFoundError,
+    MissingDatasetTypeError,
+    UnknownButlerUserError,
+)
+_USER_ERROR_MAPPING = {e.error_type: e for e in _USER_ERROR_TYPES}
+assert len(_USER_ERROR_MAPPING) == len(
+    _USER_ERROR_TYPES
+), "Subclasses of ButlerUserError must have unique 'error_type' property"
+
+
+def create_butler_user_error(error_type: str, message: str) -> ButlerUserError:
```

Comment:
Can you please add a docstring for this since it is used elsewhere and would help people reading the code and getting previews from VSCode.